### PR TITLE
Add option to output SBOM report to a file

### PR DESCRIPTION
**Pull request by wagoodman**
This PR adds a new `--file` option that writes the report bytes out to the given file path. Note: since the same functionality will be ported to grype it was decided that `-f` will not be used since that option is already taken in grype.

### CODE REVIEW

File: cmd/event_loop.go [hunk 1]
```diff
@@ -81,7 +81,7 @@ func eventLoop(workerErrs <-chan error, signals <-chan os.Signal, subscription *
 func setupUI(unsubscribe func() error, ux ui.UI) (ui.UI, error) {
 	if err := ux.Setup(unsubscribe); err != nil {
 		// replace the existing UI with a (simpler) logger UI
-		ux = ui.NewLoggerUI()
+		ux = ui.NewLoggerUI(os.Stdout)
```
Comment (luhring) [inline cmd/event_loop.go:84]: Want to make sure I follow: Why `os.Stdout` here?

File: internal/ui/select.go [hunk 2]
```diff
@@ -11,7 +12,7 @@ import (
 
 // Select is responsible for determining the specific UI function given select user option, the current platform
 // config values, and environment status (such as a TTY being present).
-func Select(verbose, quiet bool) UI {
+func Select(verbose, quiet bool, reportWriter io.Writer) UI {
```
Comment (luhring) [inline internal/ui/select.go:15]: It'd be nice to explain the correct usage of this new parameter (even very briefly) in the function's doc comment

File: internal/config/application.go [hunk 3]
```diff
@@ -29,6 +29,7 @@ type parser interface {
 type Application struct {
 	ConfigPath         string             `yaml:",omitempty" json:"configPath"`                                                         // the location where the application config was read from (either from -c or discovered while loading)
 	Output             string             `yaml:"output" json:"output" mapstructure:"output"`                                           // -o, the Presenter hint string to use for report formatting
+	File               string             `yaml:"file" json:"file" mapstructure:"file"`                                                 // -f, the file to write report output to
```
Comment (luhring) [inline internal/config/application.go:32]: We should probably remove `-f` from this line's comment for now

File: cmd/root.go [hunk 4]
```diff
@@ -40,6 +40,17 @@ func init() {
 
 	rootCmd.PersistentFlags().CountVarP(&persistentOpts.Verbosity, "verbose", "v", "increase verbosity (-v = info, -vv = debug)")
 
+	rootCmd.PersistentFlags().StringP(
+		"file", "f", "",
```
Comment (luhring) [inline cmd/root.go:44]: Should we remove the `"f"` for now?

File: cmd/report_writer.go [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+	case 0:
+		return os.Stdout, nop, nil
+	default:
+		reportFile, err := os.OpenFile(path, os.O_RDWR|os.O_CREATE|os.O_TRUNC, 0644)
+		if err != nil {
+			return nil, nop, fmt.Errorf("unable to create report file: %w", err)
+		}
+		return reportFile, func() error {
+			if !appConfig.Quiet {
+				_, _ = fmt.Fprintf(os.Stderr, "Report written to %q\n", path)
```
Comment (luhring) [inline cmd/report_writer.go:24]: Curious: Can we not use logging here?

File: cmd/root.go (hunk 4 above)
Comment (wagoodman) [reply cmd/root.go:44]: this was my bad --I removed it but was restoring work from the reflog and accidentally brought it back. Removing again...

File: cmd/event_loop.go (hunk 1 above)
Comment (luhring) [reply cmd/event_loop.go:84]: Disregard the above — in favor of this comment -> https://github.com/anchore/syft/pull/530#discussion_r722450745

Comment (wagoodman) [reply cmd/event_loop.go]: I've done a tiny refactor so that the caller can provide fallback UIs to use in case the preferred one does not work

**Review by spiffcs — APPROVED** [review]

**Review by luhring — APPROVED** [review]

**Pull request merged**